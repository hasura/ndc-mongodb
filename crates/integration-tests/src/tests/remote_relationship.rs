use crate::{connector::Connector, graphql_query, run_connector_query};
use insta::assert_yaml_snapshot;
use ndc_test_helpers::{and, asc, binop, field, query, query_request, target, variable};
use serde_json::json;

#[tokio::test]
async fn provides_source_and_target_for_remote_relationship() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This means that remote joins are not working in MongoDB 5
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query AlbumMovies($limit: Int, $movies_limit: Int) {
                  album(limit: $limit, order_by: { title: Asc }) {
                    title
                    movies(limit: $movies_limit, order_by: { title: Asc }) {
                      title
                      runtime
                    }
                    albumId
                  }
                }
            "#
        )
        .variables(json!({ "limit": 11, "movies_limit": 2 }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn handles_request_with_single_variable_set() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This means that remote joins are not working in MongoDB 5
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request()
                .collection("movies")
                .variables([[("id", json!("573a1390f29313caabcd50e5"))]])
                .query(
                    query()
                        .predicate(binop("_eq", target!("_id"), variable!(id)))
                        .fields([field!("title")]),
                ),
        )
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn variable_used_in_multiple_type_contexts() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This means that remote joins are not working in MongoDB 5
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request()
                .variables([[("dateInput", "2015-09-15T00:00Z")]])
                .collection("movies")
                .query(
                    query()
                        .predicate(and([
                            binop("_gt", target!("released"), variable!(dateInput)), // type is date
                            binop("_gt", target!("lastupdated"), variable!(dateInput)), // type is string
                        ]))
                        .order_by([asc!("_id")])
                        .limit(20)
                        .fields([
                            field!("_id"),
                            field!("title"),
                            field!("released"),
                            field!("lastupdated")
                        ]),
                )
        )
        .await?
    );
    Ok(())
}
