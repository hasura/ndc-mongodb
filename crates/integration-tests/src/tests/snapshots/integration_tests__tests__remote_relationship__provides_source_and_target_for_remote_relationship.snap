---
source: crates/integration-tests/src/tests/remote_relationship.rs
expression: "query(r#\"\n                query AlbumMovies($limit: Int, $movies_limit: Int) {\n                  album(limit: $limit, order_by: { title: Asc }) {\n                    title\n                    movies(limit: $movies_limit, order_by: { title: Asc }) {\n                      title\n                      runtime\n                    }\n                    albumId\n                  }\n                }\n            \"#).variables(json!({\n                        \"limit\": 11, \"movies_limit\": 2\n                    })).run().await?"
---
data:
  album:
    - albumId: 156
      movies:
        - runtime: 156
          title: "20th Century Boys 3: Redemption"
        - runtime: 156
          title: A Majority of One
      title: "...And Justice For All"
    - albumId: 257
      movies:
        - runtime: 257
          title: Storm of the Century
      title: "20th Century Masters - The Millennium Collection: The Best of Scorpions"
    - albumId: 296
      movies: []
      title: "A Copland Celebration, Vol. I"
    - albumId: 94
      movies:
        - runtime: 94
          title: 100 Girls
        - runtime: 94
          title: 12 and Holding
      title: A Matter of Life and Death
    - albumId: 95
      movies:
        - runtime: 95
          title: (500) Days of Summer
        - runtime: 95
          title: "1"
      title: A Real Dead One
    - albumId: 96
      movies:
        - runtime: 96
          title: "'Doc'"
        - runtime: 96
          title: "'night, Mother"
      title: A Real Live One
    - albumId: 285
      movies: []
      title: A Soprano Inspired
    - albumId: 139
      movies:
        - runtime: 139
          title: "20th Century Boys 2: The Last Hope"
        - runtime: 139
          title: 42 Up
      title: A TempestadeTempestade Ou O Livro Dos Dias
    - albumId: 203
      movies:
        - runtime: 203
          title: Michael the Brave
        - runtime: 203
          title: Michael the Brave
      title: A-Sides
    - albumId: 160
      movies:
        - runtime: 160
          title: "2001: A Space Odyssey"
        - runtime: 160
          title: 7 Aum Arivu
      title: Ace Of Spades
    - albumId: 232
      movies:
        - runtime: 232
          title: Bratya Karamazovy
        - runtime: 232
          title: Gormenghast
      title: Achtung Baby
errors: ~
